src/modules/ekf2/EKF/aid_sources/external_vision/ev_vel.h

```diff
@@ -0,0 +1,176 @@
+		Vector3f position_offset_body = _ekf._params.ev_pos_body - _ekf._params.imu_pos_body;
+		_velocity_offset_body = angular_velocity % position_offset_body;
+	}
+
+	virtual ~ExternalVisionVel() = default;
+	virtual bool fuseVelocity(estimator_aid_source3d_s &aid_src, float gate)
+	{
+		_ekf.fuseLocalFrameVelocity(aid_src, aid_src.timestamp, _measurement,
+					    _measurement_var, gate);
+		return aid_src.fused;
+
+	}
+
+	virtual void resetVelocity()
+	{
+		_ekf.resetVelocityTo(_measurement, _measurement_var);
+	}
+
+	void resetVerticalVelocity()
+	{
+		_ekf.resetVerticalVelocityTo(_measurement(2), _measurement_var(2));
+	}
+
+	void enforceMinimumVariance()
+	{
+		_min_variance = math::max(_min_variance, sq(0.01f));
+
+		for (int i = 0; i < 3; ++i) {
+			_measurement_var(i) = math::max(_measurement_var(i), _min_variance);
+		}
+	}
+
+	Ekf &_ekf;
+	const extVisionSample &_sample;
+	float _min_variance;
+	Vector3f _measurement{NAN, NAN, NAN};
+	Vector3f _measurement_var;
+	Vector3f _velocity_offset_body;
+};
+
+class EvVelBodyFrameFrd : public ExternalVisionVel
+{
+public:
+	EvVelBodyFrameFrd(Ekf &ekf_instance, extVisionSample &vision_sample, const float ev_vel_noise,
+			  const imuSample &imu_sample) :
+		ExternalVisionVel(ekf_instance, vision_sample, ev_vel_noise, imu_sample)
+	{
+		_measurement = _sample.vel - _velocity_offset_body;
+		_measurement_var = _sample.velocity_var;
+		enforceMinimumVariance();
+	}
+
+	void resetVelocity() override
+	{
+		const matrix::SquareMatrix3f rotated_variance = _ekf._R_to_earth * matrix::diag(
+					_measurement_var) * _ekf._R_to_earth.transpose();
+		// bump the variance to decrease cross-correlation and increase uncertainty of velocity
+		const Vector3f measurement_variance_ekf_frame = rotated_variance.diag() * 5.f;
+		_ekf.resetVelocityTo(_ekf._R_to_earth * _measurement, measurement_variance_ekf_frame);
+	}
+
+	void resetVerticalVelocity()
+	{
+		const matrix::SquareMatrix3f rotated_variance = _ekf._R_to_earth * matrix::diag(
+					_measurement_var) * _ekf._R_to_earth.transpose();
+		// bump the variance to decrease cross-correlation and increase uncertainty of velocity
+		const Vector3f measurement_variance_ekf_frame = rotated_variance.diag() * 5.f;
+		_ekf.resetVerticalVelocityTo((_ekf._R_to_earth * _measurement)(2, 0),
+					     measurement_variance_ekf_frame(2));
+	}
+
+	bool fuseVelocity(estimator_aid_source3d_s &aid_src, float gate) override
+	{
+		_ekf.fuseBodyFrameVelocity(aid_src, _sample.time_us, _measurement,
+					   _measurement_var, gate);
+		return aid_src.fused;
+	}
+};
+
+class EvVelLocalFrameNed : public ExternalVisionVel
+{
+public:
+	EvVelLocalFrameNed(Ekf &ekf_instance, extVisionSample &vision_sample, const float ev_vel_noise,
+			   const imuSample &imu_sample) :
+		ExternalVisionVel(ekf_instance, vision_sample, ev_vel_noise, imu_sample)
+	{
+		const Vector3f velocity_offset_earth = _ekf._R_to_earth * _velocity_offset_body;
+
+		if (_ekf._control_status.flags.yaw_align) {
```

Comment:
why can't we subtract the velocity offset if `!yaw_align`?

Reply:
When the yaw alignment between local frame and EKF frame has not converged, we should not fuse the measurement.